changefeedccl: webhook sink implementation

spiffyy99 opened the pull request:
Adds support for a new enterprise webhook-style HTTP sink that sends
JSON POST payloads for every row and resolved timestamp emitted by the
changefeed. Configurable with new options and query params as mentioned
in the corresponding RFC.

Resolves #65816

Release note (enterprise change): Introduces new HTTP sink (prefix
'webhook-https') to send individual changefeed messages as webhook
events.

RFC PR: https://github.com/cockroachdb/cockroach/pull/65927 

Epic CRDB-7748

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66497)
<!-- Reviewable:end -->

spiffyyeng commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -36,8 +36,9 @@ import (
-		`experimental-https`, `experimental-azure`:
+	case changefeedbase.SinkSchemeCloudStorageS3, changefeedbase.SinkSchemeCloudStorageGCS,
+		changefeedbase.SinkSchemeCloudStorageNodelocal, changefeedbase.SinkSchemeCloudStorageHTTP,
+		changefeedbase.SinkSchemeCloudStorageHTTPS, changefeedbase.SinkSchemeCloudStorageAzure:
```
Set these as constants for now. Will be easier to track when we change the `experimental-` prefix

spiffyyeng commented on pkg/ccl/changefeedccl/encoder.go:
```diff
@@ -99,16 +100,20 @@ func getEncoder(
+	opts map[string]string, targets jobspb.ChangefeedTargets,
+) (*jsonEncoder, error) {
 	e := &jsonEncoder{
+		targets: targets,
```
include targets to embed the name in json if needed

shermanCRL commented:
Bikeshed: we might want to use the webhook terminology throughout the code to prevent the inevitable question of how it’s different than our ‘classic’ HTTP.

miretskiy reviewed (commented):
> Reviewed 19 of 23 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_http.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUkAkCDcV3ZXeNDorR:-McUkAkCDcV3ZXeNDorS:b2lo9m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L38)):*
> ```Go
> 	applicationTypeJSON = `application/json`
> 	authorizationHeader = `Authorization`
> 	workerPoolSize      = 17
> ```

Overall, this is a good starting point.  I have few concerns and comments:
 
There is a lot of complexity around concurrent request sending (starting with this magic constant).  I think the initial version should focus
on the simplest possible implementation.  None of the concurrency; send one request at a time.  Put a todo to evaluate and implement
ability to send concurrent requests ( feel free to make a copy of this code -- you can reuse it later).

If we decide that we absolutely need to send multiple requests in parallel, then that should be addressed in separate PRs -- starting
with a PR that adds a benchmark against single threaded execution, and then adds parallelism and demonstrates appropriate performance improvements.

The "sendMessage" code does not deal with transient HTTP errors -- it should.  The current cloudstorage sink hides a lot of things it
does for us, including the retries.  To be clear: you should not implement retries in this PR.  Put a big todo.  Most likely, we'll wind up 
moving http related stuff into e.g. `webhook.go` and deal with all of the required retires (and possibly implement multi-threaded stuff there).

Finally (I mention this elsewhere), we should rename this file and the sink.  Let's use webhook throughout so that we don't
confuse with the existing http sink.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUhdog-bEEB59yivaM:-McUhdog-bEEB59yivaN:b-q1s7lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L53)):*
> ```Go
> func encodePayloadHTTP(value []byte) ([]byte, error) {
> 	var payload map[string]interface{}
> 	if err := json.Unmarshal(value, &payload); err != nil {
> ```

this looks strange to me.  Value was already encoded (with our encoder) .  We then decode it... Just so that we can encode it again.
Why do we need to do that?
___
*[pkg/ccl/changefeedccl/sink_http.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUhw1bBigvJ7kaVnjb:-McUhw1bBigvJ7kaVnjc:bmh4qwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L91)):*
> ```Go
> 		resolvedCh  chan *workerPayload
> 		flushCh     chan *httpSinkResponse
> 		wg          *sync.WaitGroup
> ```

it's strange to see so many thread safe structures (such as wg, or channels) under mutex.
If there is a reason, you should explain it.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McQziOw33OpRZ9lx5tA:-McQziOw33OpRZ9lx5tB:b-pz9mpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L95)):*
> ```Go
> }
> 
> func makeHTTPSink(ctx context.Context, u sinkURL, opts map[string]string) (Sink, error) {
> ```

should we just call it webhook? Like makeWebhookSink?
Also rename the file? 
___
*[pkg/ccl/changefeedccl/sink_http.go, line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McR-_SNDxilev2yfaGV:-McR-_SNDxilev2yfaGW:be51qsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L131)):*
<details><summary><i alt="summary">Quoted 19 lines of code…</i></summary>

> ```
> 	switch changefeedbase.FormatType(opts[changefeedbase.OptFormat]) {
> 	case changefeedbase.OptFormatJSON:
> 	// only JSON supported at this time for HTTP sink
> 	default:
> 		return nil, errors.Errorf(`this sink is incompatible with %s=%s`,
> 			changefeedbase.OptFormat, opts[changefeedbase.OptFormat])
> 	}
> 	switch changefeedbase.EnvelopeType(opts[changefeedbase.OptEnvelope]) {
> 	case changefeedbase.OptEnvelopeWrapped:
> 	default:
> 		return nil, errors.Errorf(`this sink is incompatible with %s=%s`,
> 			changefeedbase.OptEnvelope, opts[changefeedbase.OptEnvelope])
> 	}
> 	if _, ok := opts[changefeedbase.OptKeyInValue]; !ok {
> 		return nil, errors.Errorf(`this sink requires the WITH %s option`, changefeedbase.OptKeyInValue)
> 	}
> 	if _, ok := opts[changefeedbase.OptTopicInValue]; !ok {
> 		return nil, errors.Errorf(`this sink requires the WITH %s option`, changefeedbase.OptTopicInValue)
> 	}
> ```
</details>

should this validation happen before we even create sink?
___
*[pkg/ccl/changefeedccl/sink_http.go, line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McR-wnY4XjAZHr74ILK:-McR-wnY4XjAZHr74ILL:br0m91f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L237)):*
> ```Go
> 		select {
> 		case <-s.ctx.Done():
> 			close(worker)
> ```

This looks wrong to me.  You have multiple workers, so, presumably you'll close the same channel multiple times.
Furthermore, this worker reads worker payload messages -- it's a consumer; and consumers should never close themselves
(producer should).
___
*[pkg/ccl/changefeedccl/sink_http.go, line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McR0xLP3HzE_cpH1I8I:-McR0xLP3HzE_cpH1I8J:bzahhjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L259)):*
> ```Go
> // a "health check" endpoint to use.
> func (s *httpSink) Dial() error {
> 	return nil
> ```

Is is a no-op because creation of http client above doesn't actually open connection eagerly?
___
*[pkg/ccl/changefeedccl/sink_http.go, line 263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUj0k8DqEEPe8cl5Y3:-McUj0k96LdGDhJYDa-g:b-y9ci94) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L263)):*
> ```Go
> 
> func (s *httpSink) sendMessage(ctx context.Context, reqBody string) *httpSinkResponse {
> 	resWrapper := &httpSinkResponse{url: s.url}
> ```

We always include constant url in the httpsink response.  Why?
___
*[pkg/ccl/changefeedccl/sink_http.go, line 266 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUjT-F8FlIkP6qvJOm:-McUjT-F8FlIkP6qvJOn:b-b4pehh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L266)):*
> ```Go
> 
> 	var req *http.Request
> 	req, resWrapper.err = http.NewRequestWithContext(ctx, http.MethodPost, s.url.String(), strings.NewReader(reqBody))
> ```

nit: why is it a wrapper? Perhaps just response? 
___
*[pkg/ccl/changefeedccl/sink_http.go, line 284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUjfF49OPrVjXRB40c:-McUjfF49OPrVjXRB40d:bkzgpx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L284)):*
> ```Go
> 
> 	var resBody []byte
> 	resBody, resWrapper.err = ioutil.ReadAll(res.Body)
> ```

do we have to read the body of the response?
___
*[pkg/ccl/changefeedccl/sink_http_test.go, line 343 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUmEfM3d7zaYX1snZB:-McUmEfM3d7zaYX1snZC:b-rcmo2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http_test.go#L343)):*
> ```Go
> 
> 	opts := getGenericHTTPSinkOptions()
> 	opts[changefeedbase.OptHTTPClientTimeout] = "1000"
> ```

what's the unit?
___
*[pkg/ccl/changefeedccl/testfeed_test.go, line 1334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUmdx6AIFRxYvq2FS_:-McUmdx6AIFRxYvq2FSa:bdnguto) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/testfeed_test.go#L1334)):*
> ```Go
> 			m := &cdctest.TestFeedMessage{}
> 			msg := all[0]
> 			f.mockSink.RemoveFirst()
> ```

maybe drop the All method and simply have e.g. `Pop` method return and remove the first message (or "" if no messages)?
___
*[pkg/ccl/changefeedccl/cdctest/mock_http_sink.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McQyhZq2VI0vYT0XjzL:-McQyhZq2VI0vYT0XjzM:b-2xve2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/cdctest/mock_http_sink.go#L31)):*
> ```Go
> 	onRequest          func(sink *MockHTTPSink)
> 	mu                 struct {
> 		syncutil.Mutex
> ```

usually, you would include the fields protected by this mutex inside this struct.
___
*[pkg/ccl/changefeedccl/cdctest/mock_http_sink.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUnLTaBPUc0L8Iqad3:-McUnLTaBPUc0L8Iqad4:b-k7q6gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/cdctest/mock_http_sink.go#L96)):*
> ```Go
> 	s.basicAuth = true
> 	s.username = username
> 	s.password = password
> ```

does this need to happen before you start the sink?
___
*[pkg/ccl/changefeedccl/cdctest/mock_http_sink.go, line 168 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUn2OFByFcdKCtRJgS:-McUn2OFByFcdKCtRJgT:b-lkwper) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/cdctest/mock_http_sink.go#L168)):*
> ```Go
> 	s.mu.Unlock()
> 	hw.WriteHeader(s.statusCode)
> 	defer hr.Body.Close()
> ```

it's very strange to see defer like this at the very end.
Do you not want to close body unconditionally?
___
*[pkg/ccl/changefeedccl/cdctest/tls_util.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McQyOGg667sfzfBpb9N:-McQyOGg667sfzfBpb9O:b-aul6rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/cdctest/tls_util.go#L43)):*
> ```Go
> 	keyLength := 2048
> 
> 	CAKey, err := rsa.GenerateKey(rand.Reader, keyLength)
> ```

nit: `caKey`, `caCert`, etc.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUyeT-4RAwGXLVUVqi:-McUyeT-4RAwGXLVUVqj:b-aqn5sc) (no related file):*
Nice to see your progress here!  I appreciate all the effor you put into the testing code.  I haven't gotten to review it all, but I've left some initial comments that are hopefully helpful.

> None of the concurrency; send one request at a time.  Put a todo to evaluate and implement
> ability to send concurrent requests ( feel free to make a copy of this code -- you can reuse it later).

I think I agree here and apologies for leading you astray a bit since I know we chatted about the key-based hashing strategy here for handling the ordering concerns from the RFC. That said, I do think we'll want the concurrent senders eventually so hopefully that effort will be useful in a follow up PR if you do choose to remove it from here.

I've still commented on some of the concurrency-related code regardless, but feel free to ignore if the plan is to stay with a single sender for now. I've tried to mark them all with `[concurrency]`.

I think the main "missing feature" to work on would be retries.

I'll need to put some thought into the `topic_in_value` feature.  If I were writing an HTTP endpoint to consume this data, I would probably want the topic in the URL so that I could dispatch on it before reading any of the body.  But, another use-case for this endpoint is integrating with more generic systems where sometimes you want everything in the body.

---
___
*[pkg/ccl/changefeedccl/sink_http.go, line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUhdog-bEEB59yivaM:-McUuJ7s8KlPlU00s2nu:bbi36v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L53)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

this looks strange to me.  Value was already encoded (with our encoder) .  We then decode it... Just so that we can encode it again.
Why do we need to do that?
</blockquote></details>

My assumption here is that this was a way to use json.Marshal in order to produce the `payload` wrapper message around the already encoded value.

For now, I would probably go with one of two routes instead of this:

1) Manually construct the JSON wrapper it is a relatively small header we are trying to add to the message
2) Create a small struct that represents the message format (warning, code written in browser):

```go
type httpSinkPayload struct {
	Payload json.RawMessage `json:"payload"`
}
```

`json.RawMessage` is a type from the JSON package that can be used when you know your data is already well-formed JSON. (https://golang.org/pkg/encoding/json/#RawMessage)
___
*[pkg/ccl/changefeedccl/sink_http.go, line 229 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUqfFNDQpd_nFlRFf8:-McUqfFNDQpd_nFlRFf9:b-nowu0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L229)):*
> ```Go
> 	s.mu.resolvedCh = make(chan *workerPayload)
> 	s.mu.wg.Add(1)
> 	go s.workerLoop(s.mu.resolvedCh, s.mu.wg)
> ```

[concurrency] I think we'll want to look at some of the primitives that we have in the code base to avoid use of raw `go` statements.  `ctxgroup` and `stopper` provide a bit of structure around managing groups of workers.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 242 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McV1xjt4tPy7biqrGi_:-McV1xjt4tPy7biqrGia:b-olz3sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L242)):*
> ```Go
> 			message := s.sendMessage(s.ctx, req.msg)
> 			// don't expect flush for resolved timestamps
> 			if req.signalFlush {
> ```

[concurrency] I think there is probably a better pattern than spawning these short-lived goroutines.  One option would be to have a dedicated go-routine reading responses and managing the inflight-count, and doing an unconditional blocking send to that goroutine here.  Because the send is blocking, that other goroutine could then also coordinate Flush().  We can potentially separate out into a little data structure/primitive.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McV4Se12HnDjdvl34zT:-McV4Se12HnDjdvl34zU:b-24astu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L270)):*
> ```Go
> 		return resWrapper
> 	}
> 	req.Header.Set("Content-Type", applicationTypeJSON)
> ```

As a potential follow-up, I feel like the JSON format is a bit too tied up with the rest of the sink at the moment.  That isn't your fault really, it is a problem in general with the current structure of some of the sink & encoder code.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUjfF49OPrVjXRB40c:-McUpxdj5NVm-TPe9u6a:b-832vvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L284)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we have to read the body of the response?
</blockquote></details>

Depends on whether we want to reuse the connection.  To ensure re-use, you need to read to EOF and close the body.  But I'm not sure if I see a good reason to keep the body around if it isn't an error case.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 306 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUrHXhABjsXpB8f1yC:-McUrHXhABjsXpB8f1yD:bgrxatf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L306)):*
> ```Go
> 	ctx context.Context, topic TopicDescriptor, key, value []byte, updated hlc.Timestamp,
> ) error {
> 	j, err := encodePayloadHTTP(value)
> ```

If we keep the worker threads: might it make sense to push the encoding into the threads?
___
*[pkg/ccl/changefeedccl/sink_http.go, line 316 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McV3PN6CNZtLTnJ3He-:-McV3PN6CNZtLTnJ3He0:b94pszr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L316)):*
> ```Go
> 	index := workerIndex(key)
> 	payload := &workerPayload{
> 		msg:         string(j),
> ```

I would consider keeping this as []bytes throughout and then passing a [`bytes.Reader`](https://golang.org/pkg/bytes/#NewReader) to our HTTP request.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 355 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUwp988vVFysCjRVxZ:-McUwp994P7LECJsYPZn:bk3hoy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L355)):*
> ```Go
> 	var flushErr error
> 	for s.mu.inflight > 0 {
> 		log.Infof(ctx, "flush waiting for %d inflight messages", s.mu.inflight)
> ```

We'll want to move some of these logs statements to a higher verbosity level before shipping this.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McV0GyS4U2jbAaHpKOf:-McV0GyS4U2jbAaHpKOg:b810c5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L379)):*
> ```Go
> 		return err
> 	}
> 	s.client.CloseIdleConnections()
> ```

[concurrency] I think it might make sense to wait for our workers to exit before we do this.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 389 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUsBxI9fo75rgax0Ig:-McUsBxI9fo75rgax0Ih:b-69a6w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L389)):*
> ```Go
> // be the value of the HTTP header `Authorization:`. The header value will be
> // of the format `<type> <credentials>`, so the credentials will be replaced by
> // `redacted`.
> ```

This is true per the letter of the law, but I wonder if it is better to err on the safe side here and redact the entire header value just in case.
___
*[pkg/ccl/changefeedccl/changefeedbase/options.go, line 148 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUwSCb-F1lkN1b7OWP:-McUwSCb-F1lkN1b7OWQ:b2bk8lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/changefeedbase/options.go#L148)):*
> ```Go
> 	OptProtectDataFromGCOnPause: sql.KVStringOptRequireNoValue,
> 	OptKafkaSinkConfig:          sql.KVStringOptRequireValue,
> 	OptHTTPAuthHeader:           sql.KVStringOptRequireValue,
> ```

Before merge, we'll need to make a final call on whether we are going to use WITH options or query params for things like this.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_http.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUkAkCDcV3ZXeNDorR:-McjAJUiBNs5-bdFeMzl:bjcgenj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L38)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Overall, this is a good starting point.  I have few concerns and comments:
 
There is a lot of complexity around concurrent request sending (starting with this magic constant).  I think the initial version should focus
on the simplest possible implementation.  None of the concurrency; send one request at a time.  Put a todo to evaluate and implement
ability to send concurrent requests ( feel free to make a copy of this code -- you can reuse it later).

If we decide that we absolutely need to send multiple requests in parallel, then that should be addressed in separate PRs -- starting
with a PR that adds a benchmark against single threaded execution, and then adds parallelism and demonstrates appropriate performance improvements.

The "sendMessage" code does not deal with transient HTTP errors -- it should.  The current cloudstorage sink hides a lot of things it
does for us, including the retries.  To be clear: you should not implement retries in this PR.  Put a big todo.  Most likely, we'll wind up 
moving http related stuff into e.g. `webhook.go` and deal with all of the required retires (and possibly implement multi-threaded stuff there).

Finally (I mention this elsewhere), we should rename this file and the sink.  Let's use webhook throughout so that we don't
confuse with the existing http sink.

</blockquote></details>

good point! i think we weren't clear enough regarding the concurrency approach in the initial meeting. 

Let's go with a simple implementation for now. I'll keep this version on a separate branch and mention it in the RFC.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUhdog-bEEB59yivaM:-McjAs7u9nDCprLZ7vQ_:bke19uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L53)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

My assumption here is that this was a way to use json.Marshal in order to produce the `payload` wrapper message around the already encoded value.

For now, I would probably go with one of two routes instead of this:

1) Manually construct the JSON wrapper it is a relatively small header we are trying to add to the message
2) Create a small struct that represents the message format (warning, code written in browser):

```go
type httpSinkPayload struct {
	payload json.RawMessage `json:"payload"`
}
```

`json.RawMessage` is a type from the JSON package that can be used when you know your data is already well-formed JSON. (https://golang.org/pkg/encoding/json/#RawMessage)


</blockquote></details>

Yes, the intent was to have the value as a `payload` array, so we can put the batch of values in the same request (adding it initially to make the format compatible with batches in the future). I agree this needs refactoring though.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 95 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McQziOw33OpRZ9lx5tA:-McjBEb7CeKDW3YmLuel:bcmfsgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L95)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we just call it webhook? Like makeWebhookSink?
Also rename the file? 
</blockquote></details>

Yes, i'll go about renaming http -> webhook everywhere
___
*[pkg/ccl/changefeedccl/sink_http.go, line 284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUjfF49OPrVjXRB40c:-McjBWfRCzHpeaznwnnU:bksnu7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L284)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Depends on whether we want to reuse the connection.  To ensure re-use, you need to read to EOF and close the body.  But I'm not sure if I see a good reason to keep the body around if it isn't an error case.
</blockquote></details>

yeah, because we're using keep-alives this is necessary to close the body. But if we're not using it for a successful request, we can discard it (I kept it around for logging purposes but this probably isn't needed)
___
*[pkg/ccl/changefeedccl/sink_http_test.go, line 343 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUmEfM3d7zaYX1snZB:-McjC8QC4bfwRrVEWYIY:bdn0u71) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http_test.go#L343)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

what's the unit?
</blockquote></details>

i remember specifying in the RFC that it was in seconds, but we should probably switch it to a duration similar to something like `resolved`.


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -0,0 +1,305 @@
+	var payload json.RawMessage
+	if err := json.Unmarshal(value, &payload); err != nil {
+		return nil, err
+	}
```
I think we should be able to avoid this Unmarshal now that we are using json.RawMessage by casting the []byte to a json.RawMessage.

spiffyy99 reviewed (commented):
> Dismissed @miretskiy and @stevendanna from 9 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_http.go, line 263 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUj0k8DqEEPe8cl5Y3:-McoqDv_BrUv2t94wiGA:bski9vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L263)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

We always include constant url in the httpsink response.  Why?
</blockquote></details>

this was for logging purposes, but since we don't care about the response, we can remove this.
___
*[pkg/ccl/changefeedccl/sink_http.go, line 355 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUwp988vVFysCjRVxZ:-McoqMxo7BBiIxRYH-Rj:b-fmwde6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/sink_http.go#L355)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We'll want to move some of these logs statements to a higher verbosity level before shipping this.
</blockquote></details>

set as verbosity 2
___
*[pkg/ccl/changefeedccl/cdctest/mock_http_sink.go, line 168 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUn2OFByFcdKCtRJgS:-McoppToAfkMRIGRgL3R:b-ncm95x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cb64921430a767b4e590d2a88bda805a15bc68/pkg/ccl/changefeedccl/cdctest/mock_http_sink.go#L168)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

it's very strange to see defer like this at the very end.
Do you not want to close body unconditionally?
</blockquote></details>

moved to the beginning


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
@stevendanna @miretskiy thanks for the feedback so far! 

updated with the biggest changes being the renaming (http sink -> webhook sink), as well as removing the concurrent worker implementation in favour a simple one. let me know what you think.

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 205 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McyxEIW4RyQZsserq_D:-McyxEIW4RyQZsserq_E:bv77st1) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L205)):*
> ```Go
> }
> 
> func (s *webhookSink) sendMessage(ctx context.Context, reqBody []byte) error {
> ```

I think we still want to handle retries in this function.  I think that is probably what I would focus on next here.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 225 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McyxxuZ21B9sZlqBsck:-McyxxuZ21B9sZlqBscl:b-vmop11) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L225)):*
> ```Go
> 		resBody, err := ioutil.ReadAll(res.Body)
> 		if err != nil {
> 			return err
> ```

I would wrap this error so that we can still see what the original HTTP error was.  Perhaps something like `errors.Wrapf(err, "failed to read body for HTTP response with status: %s", res.Status)`
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 241 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Mcyte0YB0kAo64qh1iw:-Mcyte0YB0kAo64qh1ix:bb309ml) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L241)):*
> ```Go
> 	s.mu.Lock()
> 	s.mu.inflight++
> 	s.mu.Unlock()
> ```

Given that you've removed the concurrency implementation for now, I believe you can likely remove the machinery here for flushing.  Sink's currently are not called concurrently.

If you want to keep it to allow for future concurrent use, I would recommend wrapping this up into its own data structure. 

It seems like it could have an API pretty similar to sync.WaitGroup itself but perhaps with some logging added. Not the below is just idle musing and could probably be improved by more iteration.  

```
type inflightTracker struct {
    ...
}

func (i *inflightTracker) Add(ctx contex.Context, i int) error { ... }
func (i *inflightTracker) Done() { ... }
func (i *inflightTracker) WaitChan() chan struct{} {...}
```

such that EmitRow and flush could become:

```
func (s *webhookSink) EmitRow(
	ctx context.Context, _ TopicDescriptor, _, value []byte, _ hlc.Timestamp,
) error {
	j, err := encodePayloadWebhook(value)
	if err != nil {
		return err
	}
   	err := s.inflight.Add(ctx, 1)
        if err != nil {
              return err
        }
        defer s.inflight.Done()
        return s.sendMessage()
}

func (s *webhookSink) Flush(ctx context.Context) error {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-s.inflight.WaitChan():
			return nil
		}
}
```


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 16 files at r4, 4 of 8 files at r5, 1 of 3 files at r6, 1 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McoijgU-8fM85gOLTzE:-McoijgU-8fM85gOLTzF:b-pdnp4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bdd87962fd2dfc46396b9d2a3cc6cdc0bb8c5e9/pkg/ccl/changefeedccl/sink_webhook.go#L252)):*
<details><summary><i alt="summary">Quoted 13 lines of code…</i></summary>

> ```
> 	err = s.sendMessage(ctx, j)
> 	if err != nil {
> 		s.mu.Lock()
> 		s.mu.inflight--
> 		s.mu.Unlock()
> 		return err
> 	}
> 	go func() {
> 		select {
> 		case <-s.ctx.Done():
> 		case s.flushCh <- struct{}{}:
> 		}
> 	}()
> ```
</details>

not sure we need now.. send message is syncronous. 
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McohQT2CENcXoe13Uvn:-McohQT34-YtA8HDazA7:b-12gxlr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bdd87962fd2dfc46396b9d2a3cc6cdc0bb8c5e9/pkg/ccl/changefeedccl/sink_webhook.go#L264)):*
> ```Go
> 	}
> 
> 	return s.sendMessage(ctx, j)
> ```

don't we need to increment inflight?


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @stevendanna from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McoijgU-8fM85gOLTzE:-MczAFEa3UlXdHIGo6K1:b-us7n3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bdd87962fd2dfc46396b9d2a3cc6cdc0bb8c5e9/pkg/ccl/changefeedccl/sink_webhook.go#L252)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 	err = s.sendMessage(ctx, j)
> 	if err != nil {
> 		s.mu.Lock()
> 		s.mu.inflight--
> 		s.mu.Unlock()
> 		return err
> 	}
> 	go func() {
> 		select {
> 		case <-s.ctx.Done():
> 		case s.flushCh <- struct{}{}:
> 		}
> 	}()
> ```
not sure we need now.. send message is syncronous. 
</blockquote></details>

yeah makes sense, so i guess flush would be a no-op here then
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McohQT2CENcXoe13Uvn:-MczBbli7VeJWvJs8XlS:b-vyo6yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bdd87962fd2dfc46396b9d2a3cc6cdc0bb8c5e9/pkg/ccl/changefeedccl/sink_webhook.go#L264)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

don't we need to increment inflight?
</blockquote></details>

I thought flush wasn't called for resolved timestamps? (don't think it matters for this PR since we're taking out flush functionality for now)
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 205 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McyxEIW4RyQZsserq_D:-MczBSCL6CqmrvEXKijz:b-ku25on) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L205)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think we still want to handle retries in this function.  I think that is probably what I would focus on next here.
</blockquote></details>

yeah, I think we discussed previously that it would be a todo item though? (not in this first version of the PR)
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 241 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Mcyte0YB0kAo64qh1iw:-Mcz6oru7yCuZ06Btjlv:b1kqh90) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L241)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Given that you've removed the concurrency implementation for now, I believe you can likely remove the machinery here for flushing.  Sink's currently are not called concurrently.

If you want to keep it to allow for future concurrent use, I would recommend wrapping this up into its own data structure. 

It seems like it could have an API pretty similar to sync.WaitGroup itself but perhaps with some logging added. Not the below is just idle musing and could probably be improved by more iteration.  

```
type inflightTracker struct {
    ...
}

func (i *inflightTracker) Add(ctx contex.Context, i int) error { ... }
func (i *inflightTracker) Done() { ... }
func (i *inflightTracker) WaitChan() chan struct{} {...}
```

such that EmitRow and flush could become:

```
func (s *webhookSink) EmitRow(
	ctx context.Context, _ TopicDescriptor, _, value []byte, _ hlc.Timestamp,
) error {
	j, err := encodePayloadWebhook(value)
	if err != nil {
		return err
	}
   	err := s.inflight.Add(ctx, 1)
        if err != nil {
              return err
        }
        defer s.inflight.Done()
        return s.sendMessage()
}

func (s *webhookSink) Flush(ctx context.Context) error {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-s.inflight.WaitChan():
			return nil
		}
}
```

</blockquote></details>

these are some good ideas. i think we can keep it synchronous here, but we can reference this for future use


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 16 files at r4, 1 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUyeT-4RAwGXLVUVqi:-Md39fyp4YDx5k50h1ZT:bxgejzv) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Nice to see your progress here!  I appreciate all the effor you put into the testing code.  I haven't gotten to review it all, but I've left some initial comments that are hopefully helpful.

> None of the concurrency; send one request at a time.  Put a todo to evaluate and implement
> ability to send concurrent requests ( feel free to make a copy of this code -- you can reuse it later).

I think I agree here and apologies for leading you astray a bit since I know we chatted about the key-based hashing strategy here for handling the ordering concerns from the RFC. That said, I do think we'll want the concurrent senders eventually so hopefully that effort will be useful in a follow up PR if you do choose to remove it from here.

I've still commented on some of the concurrency-related code regardless, but feel free to ignore if the plan is to stay with a single sender for now. I've tried to mark them all with `[concurrency]`.

I think the main "missing feature" to work on would be retries.

I'll need to put some thought into the `topic_in_value` feature.  If I were writing an HTTP endpoint to consume this data, I would probably want the topic in the URL so that I could dispatch on it before reading any of the body.  But, another use-case for this endpoint is integrating with more generic systems where sometimes you want everything in the body.
</blockquote></details>

Going to give an <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments re testing.  I would just get rid of the timeout tests as well as delay functionality in the mock sink.  Perhaps @stevendanna might have more comments.

---
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 205 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McyxEIW4RyQZsserq_D:-Md38g3p5znKy_FeCF6F:b4i2cod) ([raw file](https://github.com/cockroachdb/cockroach/blob/12396786a0614475913a7a0d09dc48bc3c18b11e/pkg/ccl/changefeedccl/sink_webhook.go#L205)):*
<details><summary><i>Previously, spiffyyeng (Ryan Min) wrote…</i></summary><blockquote>

yeah, I think we discussed previously that it would be a todo item though? (not in this first version of the PR)
</blockquote></details>

Yes; todo is fine
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 197 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Md39-dd5SRwmh2mGQdr:-Md39-dd5SRwmh2mGQds:b-p7apxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59654054a367a603761ca4b7f24cee791bc9552/pkg/ccl/changefeedccl/sink_webhook_test.go#L197)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	sinkDest.SetDelay(2)
> ```

how long does this test take?
What are we trying to test here?

If we're trying to test some sort of unavailable sink, perhaps a better way would be to block the sink from 
responding altogether (for example, by arranging a channel in the test that the response handler tries to read from, and blocks).

At any rate, if this is about testing how the sink reacts to timeouts, then you simply try to return a timeout from mock handler -- without having to sleep.

I think this delay functionality should be removed from mock sink. 
___
*[pkg/ccl/changefeedccl/cdctest/mock_webhook_sink.go, line 172 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Md39XGC5TnsEg6jOvWR:-Md39XGC5TnsEg6jOvWS:b-64n4t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59654054a367a603761ca4b7f24cee791bc9552/pkg/ccl/changefeedccl/cdctest/mock_webhook_sink.go#L172)):*
> ```Go
> 		s.mu.onRequest(s)
> 	}
> 	time.Sleep(s.mu.delay)
> ```

Let's not do that.  Sleep in tests is not great.  Makes tests slow and flaky.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUyeT-4RAwGXLVUVqi:-Md3JDqB7lPPzF9R0SHx:bma3sbo) (no related file):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Going to give an :lgtm: modulo comments re testing.  I would just get rid of the timeout tests as well as delay functionality in the mock sink.  Perhaps @stevendanna might have more comments.
</blockquote></details>

:100: Thanks for bearing with us through the rounds of comments.  I think you've produced a nice base implementation on which we can iterate.  I've left two very minor nit-picks that I leave to your discretion.  My vote is to merge this after addressing Yevgeniy's test suggestions.  Let's file some issues on what we think are "must haves" before this can ship this.

---
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 105 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Md3ISDW9J8hApzU-o4D:-Md3ISDW9J8hApzU-o4E:b-ojnsf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59654054a367a603761ca4b7f24cee791bc9552/pkg/ccl/changefeedccl/sink_webhook.go#L105)):*
> ```Go
> 		var err error
> 		connTimeout, err = time.ParseDuration(timeout)
> 		if err != nil || connTimeout <= time.Duration(0) {
> ```

[nit] Combining these conditions means we lose some of the error information if it was actually a parse failure.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Md3IDBdDvjbzY3gKhlQ:-Md3IDBdDvjbzY3gKhlR:bqqmbp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59654054a367a603761ca4b7f24cee791bc9552/pkg/ccl/changefeedccl/sink_webhook.go#L109)):*
> ```Go
> 		}
> 	} else {
> 		connTimeout = 3 * time.Second
> ```

[nit] Perhaps move this to a constant `defaultConnectionTimeout`


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @stevendanna from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66497)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-McUyeT-4RAwGXLVUVqi:-Md3N0Ks9wt-s5j5ZIJu:bkgw5qw) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

:100: Thanks for bearing with us through the rounds of comments.  I think you've produced a nice base implementation on which we can iterate.  I've left two very minor nit-picks that I leave to your discretion.  My vote is to merge this after addressing Yevgeniy's test suggestions.  Let's file some issues on what we think are "must haves" before this can ship this.
</blockquote></details>

No worries, thanks for the feedback!

---
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 197 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66497#-Md39-dd5SRwmh2mGQdr:-Md3N3ur5bd0clgvG4C2:bm8acqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e59654054a367a603761ca4b7f24cee791bc9552/pkg/ccl/changefeedccl/sink_webhook_test.go#L197)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

how long does this test take?
What are we trying to test here?

If we're trying to test some sort of unavailable sink, perhaps a better way would be to block the sink from 
responding altogether (for example, by arranging a channel in the test that the response handler tries to read from, and blocks).

At any rate, if this is about testing how the sink reacts to timeouts, then you simply try to return a timeout from mock handler -- without having to sleep.

I think this delay functionality should be removed from mock sink. 
</blockquote></details>

removed all timeout-related tests and mock sink functionality


<!-- Sent from Reviewable.io -->

spiffyy99 committed:
changefeedccl: webhook sink implementation

Adds support for a new enterprise webhook-style HTTP sink that sends
JSON POST payloads for every row and resolved timestamp emitted by the
changefeed. Configurable with new options and query params as mentioned
in the corresponding RFC.

Resolves #65816

Release note (enterprise change): Introduces new webhook sink (prefix
'webhook-https') to send individual changefeed messages as webhook
events.

spiffyyeng commented:
bors r+
